dali/kernels/tuple_helpers.h

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_TUPLE_HELPERS_H_
+#define DALI_KERNELS_TUPLE_HELPERS_H_
+
+#include <tuple>
+
+namespace dali {
+
+template <typename... tuples>
+struct tuple_cat_helper;
+
+template <typename t0>
+struct tuple_cat_helper<t0> { using type = t0; };
+
+template <typename t0, typename... tuples>
+struct tuple_cat_helper<t0, tuples...>
+ : tuple_cat_helper<t0, typename tuple_cat_helper<tuples...>::type> {};
+
+template <typename... T1, typename... T2>
+struct tuple_cat_helper<std::tuple<T1...>, std::tuple<T2...>> {
+  using type = std::tuple<T1..., T2...>;
+};
+
+template <typename... tuples>
+using tuple_cat_t = typename tuple_cat_helper<tuples...>::type;
```

Comment:
delctype(std::tuple_cat(std::declval<tuples>{}...)) ?

Reply:
Yes, I know. I think this variant would build faster because compiler follows simpler rules in order to realize it. I'd have to check the real cost, though.
BTW, I remember having some problems with declval. Can't recall what that was exactly, but had to do with the fact that declval adds r-value reference.